util/component/mux.go

```diff
@@ -0,0 +1,186 @@
+// Copyright 2025 The Podseidon Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package component
+
+import (
+	"context"
+	"flag"
+	"fmt"
+
+	utilflag "github.com/kubewharf/podseidon/util/flag"
+	"github.com/kubewharf/podseidon/util/iter"
+	"github.com/kubewharf/podseidon/util/optional"
+	"github.com/kubewharf/podseidon/util/util"
+)
+
+// Provides an implementation that implements `Interface`,
+// declaring a component for the implementation and a component for the interface.
+// If the interface component is declared multiple times,
+// an implementation is selected based on the CLI options provided by the user.
+//
+// The same `implName` must not be declared twice for the same `muxName`.
+// All calls to `DeclareMuxImpl` with the same `muxName` must pass the same `Interface`.
+//
+// All options are prefixed by `{muxName}-{implName}` instead of `{implName}`.
+// Implementation components that did not get selected by the user
+// and components only requested by such components
+// do not get `init`ed, and hence their lifecycles are never invoked.
+//
+// The first implementation that got declared is always the default option.
+func DeclareMuxImpl[Args any, Options any, Deps any, State any, Interface any](
+	muxName string,
+	implNameFn func(Args) string,
+	optionsFn func(Args, *flag.FlagSet) Options,
+	depsFn func(Args, *DepRequests) Deps,
+	init func(context.Context, Args, Options, Deps) (*State, error),
+	lifecycle Lifecycle[Args, Options, Deps, State],
+	api func(*Data[Args, Options, Deps, State]) Interface,
+) func(Args) func(*DepRequests) {
+	return func(implArgs Args) func(*DepRequests) {
+		implName := implNameFn(implArgs)
+
+		implComp := Declare(
+			func(Args) string { return fmt.Sprintf("%s-%s", muxName, implName) },
```

Comment:
I am not sure if there maybe duplicate keys in the future, how about using a special split like `@`

Reply:
good idea, I'll use `.` since the name is also used as CLI option prefix.